components/script/dom/browsingcontext.rs

```diff
@@ -78,10 +103,67 @@ impl BrowsingContext {
         self.discarded.set(true);
     }
 
+    fn next_state_id(&self) -> HistoryStateId {
+        let next_id = self.next_state_id.get();
+        self.next_state_id.set(HistoryStateId(next_id.0 + 1));
+        next_id
+    }
+
+    // TODO(ConnorGBrewster): Store and do something with `title`, and `url`
+    pub fn replace_session_history_entry(&self,
+                                         title: DOMString,
+                                         url: ServoUrl,
+                                         state: HandleValue) {
+        let mut states = self.states.borrow_mut();
+        states.insert(self.active_state.get(), HistoryState::new(Some(state), Some(title), url));
+        // NOTE: We do not need to notify the constellation, as the history state id
```

Comment:
Not sure about that, what if the document is discarded then reloaded? The constellation needs to know the new URL.